### ISSUE TITLE
TrueZero Development Roadmap

### ISSUE THREAD

**Issue by andreaslam** [andreaslam/ZanLing-TrueZero#1]
# Development Roadmap:

## Monte Carlo Tree Search (MCTS):

- [x] Add and integrate settings struct 
- [x] Implement WDL and moves left 
- [x] Implement caching in MCTS
- [x] Optimise caching performance
- [x]  Refactor file format system
- [x] Implement root FPU
- [ ] Implement virtual loss in MCTS
- [ ] Add draw weight and adjustable aggressiveness
- [ ] Implement support for Endgame Tablebases
- [ ] Implement playout cap randomisation
- [ ] Research on MCGS

## Data Generation Infrastructure:
- [x] Support multi-GPU data generation

## Training Infrastructure:

- [x] Create file buffer
- [x] Create basic training loop
- [x] Update struct format in `SearchSettings`
- [x] TCP server client
- [ ] Support multi-GPU training

### TrueScheduler
- [ ] Implement home screen
- [ ] Implement experiment control centre to monitor and control training hyperparameters 

### TensorBoard integration
- [ ] Implement backend for server connection

## Testing and UCI support
- [x] Create UCI support for benchmarking and testing
- [x] Implement `setoptions` for UCI display
- [x] Rewrite game adjudication logic for `enginetest.rs`
- [x] Create utility toolkit for debugging
- [ ] Implement `currmove` for UCI display
- [ ] Implement refined time management system
- [ ] Implement multi-GPU inference and make it available in UCI

### V1 Server and Communication System
- [x] Facilitate communication between Python `client.py`, Rust server and data generation code
- [x] Create system to intelligently log in, verify and confirm identity by providing a unique identifier for each client joined in server
- [x] Implement system to distinguish between different types of requests and redirect to appropriate recipient if necessary 

### V2 Server and Communication System

- [x] Integrate `MessageSend` and `MessageRecv` Enums
- [x] V2 Message Parsing System
- [x] Remove message length-dependant buffers and checks
- [x] Use JSON instead of string formatting
- [x] `BufReader` implementation

### V3 Server and Communication System
- [x] Receive game data in raw bytes
- [x] Implement dynamic saving and deleting games to save disk space
- [x] Send neural network data in raw bytes
- [x] Send game data in raw bytes
- [x] Implement dynamic saving and deleting games to save disk space
- [x] Receive neural network data in raw bytes

### V4 Server and Communication System

- [x] Finalise `purpose` using Enums instead of `String`s 
- [x] Rework the messaging system to send data within an Enum inside JSON
- [x] Implement Enum checking instead of directly using `String` literals

### V5 Server and Communication System

- [x] Refine disconnection logic and reduce resource usage
- [ ] Refine message parsing system in Python (PyO3 integration)
- [ ] Enable remote controlling and scheduling settings in a GUI instead of hardcoding hyperparameters

### TrueNet
- [x] Implement learning rate schedulers
- [ ] Implement architecture features eg MaxPooling, Dropout layers
- [ ] Enable mixed-precision training
- [ ] Assess the possibility of reducing scalar input sizes (board input representation)
- [ ] Assess the possibility of using various Machine Learning techniques to accelerate convergence, eg cosine annealing and gradient accumulation
- [ ] Assess the possibility of ensemble training
- [ ] Implement TrueNet XS, a smaller neural network which initialises as flat policy and used primarily as "from scratch, zero human knowledge" synthetic data. This net will be gradually trained on the Python training buffer. This will serve as the first generation distilled network. 
- [ ] Research on decision transformers

### Reinforcement Learning (TBC):
- [ ] [Read knowledge distillation paper](https://arxiv.org/pdf/2106.05237.pdf)
- [ ] [Read KataGo paper](https://arxiv.org/pdf/1902.10565.pdf)
